activerecord/lib/active_record/insert_all.rb

```diff
@@ -0,0 +1,155 @@
+
+      ensure_valid_options_for_connection!
+    end
+
+    def execute
+      if inserts.present?
+        connection.exec_query to_sql, "Bulk Insert"
+      else
+        ActiveRecord::Result.new([], [])
+      end
+    end
+
+    def updatable_columns
+      keys - readonly_columns - unique_by_columns
+    end
+
+    private
+      def ensure_valid_options_for_connection!
+        if returning && !connection.supports_insert_returning?
+          raise ArgumentError, "#{connection.class} does not support :returning"
+        end
+
+        unless %i{ raise skip update }.member?(on_duplicate)
+          raise NotImplementedError, "#{on_duplicate.inspect} is an unknown value for :on_duplicate. Valid values are :raise, :skip, and :update"
+        end
+
+        if on_duplicate == :skip && !connection.supports_insert_on_duplicate_skip?
+          raise ArgumentError, "#{connection.class} does not support skipping duplicates"
+        end
+
+        if on_duplicate == :update && !connection.supports_insert_on_duplicate_update?
+          raise ArgumentError, "#{connection.class} does not support upsert"
+        end
+
+        if unique_by && !connection.supports_insert_conflict_target?
+          raise ArgumentError, "#{connection.class} does not support :unique_by"
+        end
+      end
+
+      def to_sql
+        connection.build_insert_sql(ActiveRecord::InsertAll::Builder.new(self))
+      end
+
+      def keys
+        inserts.present? ? inserts.first.keys.map(&:to_s) : []
+      end
+
+      def readonly_columns
+        primary_keys + model.readonly_attributes.to_a
+      end
+
+      def unique_by_columns
+        unique_by ? unique_by.fetch(:columns).map(&:to_s) : []
+      end
+
+      def primary_keys
+        Array.wrap(model.primary_key)
+      end
+
+
+      class Builder
+        attr_reader :model
+
+        def initialize(insert_all)
+          @insert_all, @model, @connection = insert_all, insert_all.model, insert_all.connection
+        end
+
+        def skip_duplicates?
+          insert_all.on_duplicate == :skip
+        end
+
+        def update_duplicates?
+          insert_all.on_duplicate == :update
+        end
+
+        def into
+          [ "INTO ", model.quoted_table_name, "(", columns_list, ") " ].join
+        end
+
+        def values_list
+          columns = connection.schema_cache.columns_hash(model.table_name)
+
+          values_list = insert_all.inserts.map do |attributes|
+            attributes = attributes.stringify_keys
+
+            unless attributes.keys.to_set == keys.to_set
+              raise ArgumentError, "All objects being inserted must have the same keys"
+            end
+
+            keys.map do |name|
+              column = columns[name]
+              type = connection.lookup_cast_type_from_column(column)
+              bind = Relation::QueryAttribute.new(name, attributes[name], type)
+              connection.with_yaml_fallback(bind.value_for_database)
```

Comment:
Is this right?  If a column is supposed to serialize as JSON, does `bind.value_for_database` return a string, or a complex object?  This could be correct, I'm just not clear why we need  it.

Reply:
Haha. Well, I don't know if I can answer that. I was just copying what `insert_fixture` does [here](https://github.com/rails/rails/blob/v6.0.0.beta2/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L349-L359) and what `build_fixture_sql` does [here](https://github.com/rails/rails/blob/v6.0.0.beta2/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L425-L427).

According to git blame, `with_yaml_fallback` was introduced in 4ee42379cc30a07a7d47b7be8c710dba0efa407a (where it was extracted) and the idea itself was probably introduced in aa31d21f5f4fc4d679e74a60f9df9706da7de373.

It _does_ get exercised by ActiveRecord's test suite — when we load [this fixture](https://github.com/rails/rails/blob/v6.0.0.beta2/activerecord/test/fixtures/admin/users.yml#L8-L10).

I think it's purpose is: if the column is a string type but the value is an array or hash, YAML it.

And I figure it makes sense to keep this behavior in InsertAll because that's generally how ActiveRecord handles arrays and hashes passed to string fields, isn't it?